skymp5-server/cpp/lib_espm/espm.cpp

```diff
@@ -1074,21 +1074,31 @@ espm::Effects::Data espm::Effects::GetData(
   }
 
   Data result;
-  uint32_t effect = 0;
+  uint32_t effectIndex = 0;
+  bool orderFlag = true;
+  bool isValid = true;
 
   espm::RecordHeaderAccess::IterateFields(
     parent,
     [&](const char* type, uint32_t size, const char* data) {
       if (!memcmp(type, "EFID", 4)) {
+        isValid = orderFlag == true;
         result.effects.emplace_back();
-        result.effects[effect].effectId =
+        result.effects[effectIndex].effectId =
           *reinterpret_cast<const uint32_t*>(data);
+        orderFlag = false;
       } else if (!memcmp(type, "EFIT", 4)) {
-        Effect& eff = result.effects[effect];
+        isValid = orderFlag == false;
+        Effect& eff = result.effects[effectIndex];
         eff.magnitude = *reinterpret_cast<const float*>(data);
         eff.areaOfEffect = *reinterpret_cast<const uint32_t*>(data + 4);
         eff.duration = *reinterpret_cast<const uint32_t*>(data + 8);
-        effect++;
+        effectIndex++;
+        orderFlag = true;
+      }
+      if (!isValid) {
+        auto name = parent->GetEditorId(compressedFieldsCache);
+        throw std::runtime_error("Bad effect array. " + *name);
```

Comment:
Hmm, BTW this will result in an empty string if for whatever reason there's no EDID. Or is this not possible? Anyway let's add form id if it's possible and not too hard

Reply:
very unlikely. vanilla and sweetpie effects always have EDID.

Reply:
but ok let's output form id if there is no EDID...
we will be able to reuse this pattern in future features.

Reply:
UPD: Let's leave it as it is for now